packages/grafana-prometheus/src/querybuilder/components/MetricCombobox.tsx

```diff
@@ -0,0 +1,125 @@
+import { useCallback } from 'react';
+
+import { SelectableValue } from '@grafana/data';
+import { EditorField, EditorFieldGroup } from '@grafana/experimental';
+import { InlineField, InlineFieldRow } from '@grafana/ui';
+import { Combobox, ComboboxOption } from '@grafana/ui/src/components/Combobox/Combobox';
+
+import { PrometheusDatasource } from '../../datasource';
+import { regexifyLabelValuesQueryString } from '../parsingUtils';
+import { QueryBuilderLabelFilter } from '../shared/types';
+import { PromVisualQuery } from '../types';
+
+export interface MetricComboboxProps {
+  metricLookupDisabled: boolean;
+  query: PromVisualQuery;
+  onChange: (query: PromVisualQuery) => void;
+  onGetMetrics: () => Promise<SelectableValue[]>;
+  datasource: PrometheusDatasource;
+  labelsFilters: QueryBuilderLabelFilter[];
+  onBlur?: () => void;
+  variableEditor?: boolean;
+}
+
+export function MetricCombobox({
+  datasource,
+  query,
+  onChange,
+  onGetMetrics,
+  labelsFilters,
+  variableEditor,
+}: Readonly<MetricComboboxProps>) {
+  /**
+   * Gets label_values response from prometheus API for current autocomplete query string and any existing labels filters
+   */
+  const getMetricLabels = useCallback(
+    async (query: string) => {
+      const results = await datasource.metricFindQuery(formatKeyValueStringsForLabelValuesQuery(query, labelsFilters));
+
+      const resultsOptions = results.map((result) => {
+        return {
+          label: result.text,
+          value: result.text,
+        };
+      });
+      return resultsOptions;
+    },
+    [datasource, labelsFilters]
+  );
+
+  const onComboboxChange = useCallback(
+    (opt: ComboboxOption<string> | null) => {
+      onChange({ ...query, metric: opt?.value ?? '' });
+    },
+    [onChange, query]
+  );
+
+  const loadOptions = useCallback(
+    async (input: string): Promise<ComboboxOption[]> => {
+      const metrics = input.length ? await getMetricLabels(input) : await onGetMetrics();
```

Comment:
When a user types input, then blurs, then returns to the combobox, the input is "" even if there is text in the input. This results in calling `onGetMetrics` and retrieves all the metrics while there is visible text in the combobox. Is this a combobox bug for the onChange callback?This feels weird when there is text entered to filter the results, but then no filtering occurs and we get all the metrics. 🤔 

![Screenshot 2024-10-28 at 4 19 03 PM](https://github.com/user-attachments/assets/2f677c95-de70-428d-abcc-f321e030f7d4)

I would suggest storing the input in state and checking that instead of the combobox input before making the async call for metrics or maybe combobox has a setting to handle this?

Reply:
Yeah, this is a Combobox bug 

https://github.com/grafana/grafana/blob/6c0d09cdb1d8f4132a5ec8bbaec57a247adbe98b/packages/grafana-ui/src/components/Combobox/Combobox.tsx#L198-L209

Current async behaviour calls loadOptions too many times at the moment, which is something we need to sort out (it also calls loadOptions after selecting an item and closing the menu)

Reply:
https://github.com/grafana/grafana/pull/95573